### DISCUSSION TITLE
Custom notifier channel.

### DISCUSSION THREAD

**Discussion by berkut1** [symfony/symfony#47579] [Q&A]
Hello,

Not so long ago I was wondering how to make notifications through a database like facebook/github/etc. https://github.com/symfony/symfony/discussions/46933

But now I'm stuck with a problem. How can I create my own channel for this, for example "database" or "doctrine"?
Thanks.

**Comment by berkut1** [comment]
I checked some files like this https://github.com/symfony/framework-bundle/blob/6.1/Resources/config/notifier.php
It looks like it's not possible to add a custom notification channel via *.yaml.

For now I made this solution (decorated the Notifier), I don't know if it's correct, but it works the way I want.

```
final class Notifier
{
    private NotifierInterface $notifier;
    private ChannelInterface $channel;

    public function __construct(NotifierInterface $notifier, #[Autowire(service: DoctrineChannel::class)] ChannelInterface $channel)
    {
        $this->notifier = $notifier;
        $this->channel = $channel;
    }

    public function send(Notification $notification, RecipientInterface ...$recipients): void
    {
        foreach ($recipients as $key => $recipient) {
            foreach ($notification->getChannels($recipient) as $channel){
                if ($channel === DoctrineChannel::name() && $this->channel->supports($notification, $recipient)) {
                    $this->channel->notify($notification, $recipient);
                    unset($recipients[$key]); //remove a used doctrine recipient
                }
            }
        }
        $recipients = array_values($recipients);

        $channels = array_diff($notification->getChannels(new NoRecipient()), [DoctrineChannel::name()]); //delete if doctrine channel exists
        if (count($channels) > 0) {
            $notification->channels($channels);
            $this->notifier->send($notification, ...$recipients); //send what left to the native notifier
        }
    }
}
```

**Reply by adrolter** [reply]
You were just missing a service tag (and optionally an alias). 

For example, using attributes:

```php
#[AsAlias('notifier.channel.doctrine')]
#[AutoconfigureTag('notifier.channel', ['channel' => 'doctrine'])]
class DoctrineChannel {
...
```

See: [symfony/framework-bundle/Resources/config/notifier.php](https://github.com/symfony/symfony/blob/6c3d3770c166711a9440778ae43f06999c668292/src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier.php#L50)
